pydantic_ai_slim/pydantic_ai/models/xai.py

```diff
@@ -0,0 +1,1046 @@
+                        web_search_tool_call = builtin_call
+                    elif item.tool_name.startswith(MCPServerTool.kind):
+                        mcp_tool_call = builtin_call
+
+            elif isinstance(item, BuiltinToolReturnPart):
+                # Update tool call status based on return part
+                self._update_builtin_tool_status(item, code_execution_tool_call, web_search_tool_call, mcp_tool_call)
+            elif isinstance(item, FilePart):  # pragma: no cover
+                # Files generated by models (e.g., from CodeExecutionTool) are not sent back
+                pass
+            else:
+                assert_never(item)
+
+        # Create assistant message with content, reasoning_contents, encrypted_contents and tool_calls
+        return self._build_assistant_message(texts, reasoning_contents, encrypted_contents, tool_calls)
+
+    def _map_tool_call(self, tool_call_part: ToolCallPart) -> chat_types.chat_pb2.ToolCall:
+        """Map a ToolCallPart to an xAI SDK ToolCall."""
+        return chat_types.chat_pb2.ToolCall(
+            id=tool_call_part.tool_call_id,
+            function=chat_types.chat_pb2.FunctionCall(
+                name=tool_call_part.tool_name,
+                arguments=tool_call_part.args_as_json_str(),
+            ),
+        )
+
+    def _map_builtin_tool_call_part(self, item: BuiltinToolCallPart) -> chat_types.chat_pb2.ToolCall | None:
+        """Map a BuiltinToolCallPart to an xAI SDK ToolCall with appropriate type and status."""
+        if not item.tool_call_id:
+            return None
+
+        if item.tool_name == CodeExecutionTool.kind:
+            return chat_types.chat_pb2.ToolCall(
+                id=item.tool_call_id,
+                type=chat_types.chat_pb2.TOOL_CALL_TYPE_CODE_EXECUTION_TOOL,
+                status=chat_types.chat_pb2.TOOL_CALL_STATUS_COMPLETED,
+                function=chat_types.chat_pb2.FunctionCall(
+                    name=CodeExecutionTool.kind,
+                    arguments=item.args_as_json_str(),
+                ),
+            )
+        elif item.tool_name == WebSearchTool.kind:
+            return chat_types.chat_pb2.ToolCall(
+                id=item.tool_call_id,
+                type=chat_types.chat_pb2.TOOL_CALL_TYPE_WEB_SEARCH_TOOL,
+                status=chat_types.chat_pb2.TOOL_CALL_STATUS_COMPLETED,
+                function=chat_types.chat_pb2.FunctionCall(
+                    name=WebSearchTool.kind,
+                    arguments=item.args_as_json_str(),
+                ),
+            )
+        elif item.tool_name.startswith(MCPServerTool.kind):
+            return chat_types.chat_pb2.ToolCall(
+                id=item.tool_call_id,
+                type=chat_types.chat_pb2.TOOL_CALL_TYPE_MCP_TOOL,
+                status=chat_types.chat_pb2.TOOL_CALL_STATUS_COMPLETED,
+                function=chat_types.chat_pb2.FunctionCall(
+                    name=item.tool_name,
```

Comment:
Is this correct? The server doesn't expect some specific tool name?

Reply:
No, we use the tool type to determine if we should call an MCP function or not.